### PR TITLE
Support structure type local variable initialization

### PR DESCRIPTION
**Pull request by splasky**
This commit fix the following things:
1. Local variable declaration have bug after expression.
2. Struct type of pointer can be initialize inside local area.
3. Update context error message.

### CODE REVIEW

File: amacc.c [hunk 1]
```diff
@@ -871,73 +871,61 @@ void stmt(int ctx)
...
-                    if (tk != Id) fatal("bad struct member definition");
-                    m = malloc(sizeof(struct member_s));
-                    m->id = id;
-                    m->offset = i;
-                    m->type = ty;
-                    m->next = members[bt];
-                    members[bt] = m;
-                    i += (ty >= PTR) ? sizeof(int) : tsize[ty];
-                    i = (i + 3) & -4;
+        switch(tk){
```
Comment (jserv) [inline amacc.c]: Be aware of the consistent style. Should be `switch (tk) {`

File: amacc.c [hunk 2]
```diff
@@ -871,73 +871,61 @@ void stmt(int ctx)
...
+        case Struct:
+                next();
+                if (tk == Id) {
```
Comment (jserv) [inline amacc.c]: Indent here.

File: amacc.c [hunk 3]
```diff
@@ -946,8 +934,16 @@ void stmt(int ctx)
             // if the beginning of * is a pointer type, then type plus `PTR`
             // indicates what kind of pointer
             while (tk == Mul) { next(); ty += PTR; }
-            if (tk != Id) fatal("bad global declaration");
-            if (id->class >= ctx) fatal("duplicate global definition");
+            switch(ctx){
```
Comment (jserv) [inline amacc.c]: Ditto. Check the style.

File: amacc.c [hunk 4]
```diff
@@ -946,8 +934,16 @@ void stmt(int ctx)
             // if the beginning of * is a pointer type, then type plus `PTR`
             // indicates what kind of pointer
             while (tk == Mul) { next(); ty += PTR; }
-            if (tk != Id) fatal("bad global declaration");
-            if (id->class >= ctx) fatal("duplicate global definition");
+            switch (ctx) {
+                case Glo:
```
Comment (jserv) [inline amacc.c]: Indent here.

File: amacc.c [hunk 5]
```diff
@@ -946,8 +934,16 @@ void stmt(int ctx)
...
+                    if(tk != Id) fatal("bad global declaration");
+                    if (id->class >= ctx) fatal("duplicate global definition");
+                    break;
+                case Loc:
```
Comment (jserv) [inline amacc.c]: Ditto.

File: amacc.c [hunk 6]
```diff
@@ -965,7 +961,7 @@ void stmt(int ctx)
                 // e represents the address which will store pc
                 // (ld - loc) indicates memory size to allocate
                 int *t = n;
-                *--n = ';'; while (tk != '}') { t = n;  stmt(Loc); *--n = (int) t; *--n = '{'; }
+                *--n = ';'; while (tk != '}') { t = n; stmt(Loc); if (t != n){*--n = (int) t; *--n = '{';} }
```
Comment (jserv) [inline amacc.c]: Split these statements into several lines for readability.

File: amacc.c [hunk 7]
```diff
@@ -871,73 +871,61 @@ void stmt(int ctx)
...
+                            while (tk == Mul) { next(); ty += PTR; }
+                            if (tk != Id) fatal("bad struct member definition");
+                            m = malloc(sizeof(struct member_s));
+                            m->id = id;
+                            m->offset = i;
+                            m->type = ty;
+                            m->next = members[bt];
+                            members[bt] = m;
+                            i += (ty >= PTR) ? sizeof(int) : tsize[ty];
+                            i = (i + 3) & -4;
```
Comment (jserv) [inline amacc.c]: I'm wondering if we shall prepare helper functions for alignment.

Comment (splasky) [reply amacc.c]: Mmm. But autoformat tool will increase the code line,won't it?

Comment (jserv) [reply amacc.c]: After code cleanup, I would like to enforce `clang-format` for consistent style. Sooner or later.

File: amacc.c [hunk 8]
```diff
@@ -871,73 +871,61 @@ void stmt(int ctx)
...
-                case Int: next(); break;
-                case Char: next(); mbt = CHAR; break;
-                case Struct:
-                    next(); 
-                    if (tk != Id) fatal("bad struct declaration");
-                    mbt = id->stype;
-                    next(); break;
+        switch (tk) {
+        case Struct:
+                next();
```
Comment (jserv) [inline amacc.c]: 4-space indention.

File: tests/local.c [hunk 9]
```diff
@@ -7,5 +12,10 @@ int main(void)
     char cc = 'a';
     char *ptr = &cc;
     printf("%c\n", *ptr);
+    struct Value a;
```
Comment (jserv) [inline tests/local.c]: Shall we add new entries about pointers to structures for `struct` test item?

Comment (splasky) [reply tests/local.c]: Like put into the `tests/struct.c`?

Comment (jserv) [reply tests/local.c]: It would be better if all structure related test entries are landed in file `tests/struct.c`

File: amacc.c [hunk 10]
```diff
@@ -946,8 +934,16 @@ void stmt(int ctx)
             // if the beginning of * is a pointer type, then type plus `PTR`
             // indicates what kind of pointer
             while (tk == Mul) { next(); ty += PTR; }
-            if (tk != Id) fatal("bad global declaration");
-            if (id->class >= ctx) fatal("duplicate global definition");
+            switch (ctx) {
+            case Glo:
+                if(tk != Id) fatal("bad global declaration");
```
Comment (jserv) [inline amacc.c]: Should be `if (tk != Id)` for consistent style.

File: amacc.c [hunk 11]
```diff
@@ -946,8 +934,16 @@ void stmt(int ctx)
...
+                if (id->class >= ctx) fatal("duplicate global definition");
+                break;
+            case Loc:
+                if(tk != Id) fatal("bad local declaration");
```
Comment (jserv) [inline amacc.c]: Ditto.

File: amacc.c [hunk 12]
```diff
@@ -871,73 +871,61 @@ void stmt(int ctx)
...
+                        case Char: next(); mbt = CHAR; break;
+                        case Struct:
+                            next(); 
+                            if (tk != Id) fatal("bad struct declaration");
+                            mbt = id->stype;
+                            next(); break;
+                        }
+                        while (tk != ';') {
+                            ty = mbt;
+                            while (tk == Mul) { next(); ty += PTR; }
```
Comment (jserv) [inline amacc.c]: Add comments back as following:
```c
// if the beginning of * is a pointer type, then type plus `PTR`
// indicates what kind of pointer
```

**Pull request merged**